src/main/java/ru/vk/itmo/test/nikitaprokopev/MyServer.java

```diff
@@ -68,117 +67,180 @@ private static HttpServerConfig createServerConfig(ServiceConfig serviceConfig)
-            return proxyRequest(Request.METHOD_GET, id, targetNode, Response.EMPTY);
+        if (request.getHeader(HEADER_INTERNAL) != null) {
+            Response response = handleInternalRequest(request);
+            try {
+                session.sendResponse(response);
+            } catch (IOException e) {
+                log.error("Exception while sending response", e);
+                session.scheduleClose();
+            }
+            return;
         }
 
-        MemorySegment msKey = toMemorySegment(id);
-
-        Entry<MemorySegment> entry = dao.get(msKey);
-
-        if (entry == null) {
-            return new Response(Response.NOT_FOUND, Response.EMPTY);
+        List<String> targetNodes = getNodesSortedByRendezvousHashing(key, serviceConfig, from);
+        List<HttpRequest> httpRequests =
+                requestHandler.createRequests(request, key, targetNodes, serviceConfig);
+        List<Response> responses = getResponses(request, ack, httpRequests);
+        if (responses.isEmpty() || responses.size() < ack) {
+            sendResponseWithEmptyBody(session, CustomResponseCodes.RESPONSE_NOT_ENOUGH_REPLICAS.getCode());
+            return;
         }
+        try {
+            if (request.getMethod() != Request.METHOD_GET) {
+                session.sendResponse(responses.getFirst());
+                return;
+            }
 
-        return Response.ok(toByteArray(entry.value()));
+            mergeGetResponses(session, responses);
+        } catch (IOException e) {
+            logIOExceptionAndCloseSession(session, e);
+        }
     }
 
-    @Path(BASE_PATH)
-    @RequestMethod(Request.METHOD_DELETE)
-    public Response delete(@Param(value = "id", required = true) String id) throws IOException {
-        if (id.isEmpty()) {
-            return new Response(Response.BAD_REQUEST, Response.EMPTY);
+    private void mergeGetResponses(HttpSession session, List<Response> responses) throws IOException {
+        byte[] body = null;
+        int notFoundResponsesCount = 0;
+        long maxTimestamp = Long.MIN_VALUE;
+        for (Response response : responses) {
+            long timestamp = response.getHeader(HEADER_TIMESTAMP) == null ? -1
+                    : Long.parseLong(response.getHeader(HEADER_TIMESTAMP));
+            if (response.getStatus() == NOT_FOUND_CODE) {
+                notFoundResponsesCount++;
+                if (timestamp != -1 && maxTimestamp < timestamp) {
+                    maxTimestamp = timestamp;
+                    body = null;
+                }
+                continue;
+            }
+            if (maxTimestamp < timestamp) {
+                maxTimestamp = timestamp;
+                body = response.getBody();
+            }
         }
 
-        int targetNode = hash(id);
-        if (targetNode != nodeId) {
-            return proxyRequest(Request.METHOD_DELETE, id, targetNode, Response.EMPTY);
+        if (body != null && notFoundResponsesCount != responses.size()) {
```

Comment:
Зачем второе условие? Разве может быть `body` и при этом все ответы 404?
```suggestion
        if (body != null) {
```

Reply:
И нужен ли вообще тогда счётчик `notFoundResponsesCount`?

Reply:
Исправил. Убрал, не нужен